modules: mbed TLS: Bump to 3.0

d3zd3z opened the pull request:
Continuation of #37753. Migrating Mbed TLS to 3.0.

Commit:
modules: mbdetls: define MBEDTLS_SHA224_c

Newer versions of mbedTLS require MBEDTLS_SHA224_C definition to use
sha256.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: mbedtls: Follow changes in mbedTLS 3.0 version

Apply the modifications required by the newer library version.
These modifications came from the original file in the mbedTLS
repository.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: crypto: Follow changes in mbedTLS 3.0 version

Apply the modifications required by the newer library version.
These modifications came from the original file in the mbedTLS
repository.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
sockets: tls: Enable access to mbedtls private fields

Several fields of structures in mbedTLS 3.0 are now private. To access
them directly is necessary to define MBEDTLS_ALLOW_PRIVATE_ACCESS.

That is a temporary fix, the proper solution is not access directly
but using proper API.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
modules: mbedTLS: Bump mbedTLS to 3.0

Bump mbedTLS version.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
net: tcp2: Fix mbedTLS usage

s/mbedtls_md5_ret/mbedtls_md5/

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
random: Fix sys_csrand_get build condition

z_impl_sys_csrand_get is implement if the system is build with either
CONFIG_CTR_DRBG_CSPRNG_GENERATOR or CONFIG_HARDWARE_DEVICE_CS_GENERATOR.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
jwt: mbedTLS requires csprng

Newer version of mbedTLS requires a csprng source now in some APIs.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: jwt: This test requires csprng

csprng is required to use JWT.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
net: sockets_tls: Fix mbedTLS usage

mbedtls_pk_parse_key signature has changed and requires an entropy
source.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
net: websocket: s/mbedtls_sha1_ret/mbedtls_sha1/

sha1 function signature changed in mbedTLS 3.0.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

d3zd3z committed:
samples: tfm_integration: Fix PSA crypto APIs

Moving to Mbed TLS 3.0 changes some of the Mbed TLS to better match the
PSA spec.  Fix up the things where we are affected by these API changes.

Signed-off-by: David Brown <david.brown@linaro.org>

carlescufi commented:
@SebastianBoe FYI

SebastianBoe commented:
So you want to bump mbedTLS, mcuboot, and TF-M simultaneously ?

I guess that could make sense if it is difficult to make mbedtls 3.0.0 work with the current versions of TF-M and mcuboot.

Or is this just an experiment?

d3zd3z commented:
> So you want to bump mbedTLS, mcuboot, and TF-M simultaneously ?
>
> I guess that could make sense if it is difficult to make mbedtls 3.0.0 work with the current versions of TF-M and mcuboot.
>
> Or is this just an experiment?

I'm not sure it is worth the trickery of trying to make some of them work with mbed TLS 2.x and some with 3.x. The change I just pushed moves them all together. The mcuboot change should continue to work with mbed TLS 2.x, but I'm not sure if the others will. There are quite a few API changes.

I just pushed the update to mbed TLS as well, so all should move together with this.

d3zd3z commented:
@SebastianBoe Presumably we will need #37932 as well, right?

microbuilder commented:
Tested with `mps2_an521_ns` and works, but `lpcxpesso55s69_ns` requires some changes to https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/60 which I'll add there. This resolves the TF-M + Mbed TLS 3.0.0 build issues, though.

d3zd3z committed:
net: sockets: tls: Use better error code

Mbed TLS 3.0 removes the definition for MBED_ERR_SSL_PEER_VERIFY_FAILED,
since non of its code ever returns that value.  Since there isn't really
a perfect response, instead return a somewhat generic response
indicating this was unexpected.

Signed-off-by: David Brown <david.brown@linaro.org>

nashif commented:
can you rebase that to get a clean CI run?

microbuilder commented:
@d3zd3z I pushed out a change to the related TF-M PR that should fix the issues with the lpcxpresso55s69.

SebastianBoe commented:
> @SebastianBoe Presumably we will need #37932 as well, right?

Someone will need #37932 at some point, but it is not critical, we can bump the revisions first.

SebastianBoe committed:
tfm: Fix zephyr-sdk toolchain variant integration

The psa-arch-tests project changed how third-party toolchains are
integrated. This broke the Zephyr sdk.

This patch aligns with the new mechanism.

https://github.com/ARM-software/psa-arch-tests/pull/276

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
Seems the mbedtls bump broke openthread. From CI:



```
FAILED: modules/openthread/build/src/core/CMakeFiles/openthread-ftd.dir/common/random_manager.cpp.obj
--
  | ccache /opt/toolchains/zephyr-sdk-0.13.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-g++ -DBUILD_VERSION=zephyr-v2.6.0-2909-ge0578a257278 -DKERNEL -DMBEDTLS_CONFIG_FILE=\"config-tls-generic.h\" -DNRF52840_XXAA -DNRF_802154_ACK_TIMEOUT_ENABLED=1 -DNRF_802154_CCA_CORR_LIMIT_DEFAULT=2 -DNRF_802154_CCA_CORR_THRESHOLD_DEFAULT=45 -DNRF_802154_CCA_ED_THRESHOLD_DEFAULT=45 -DNRF_802154_CCA_MODE_DEFAULT=NRF_RADIO_CCA_MODE_ED -DNRF_802154_CSMA_CA_ENABLED=1 -DNRF_802154_DELAYED_TRX_ENABLED=0 -DNRF_802154_ENCRYPTION_ENABLED=0 -DNRF_802154_FRAME_TIMESTAMP_ENABLED=0 -DNRF_802154_IE_WRITER_ENABLED=0 -DNRF_802154_IFS_ENABLED=0 -DNRF_802154_INTERNAL_RADIO_IRQ_HANDLING=0 -DNRF_802154_PENDING_EXTENDED_ADDRESSES=16 -DNRF_802154_PENDING_SHORT_ADDRESSES=16 -DNRF_802154_RX_BUFFERS=16 -DNRF_802154_SECURITY_WRITER_ENABLED=0 -DNRF_802154_TX_STARTED_NOTIFY_ENABLED=1 -DNRF_802154_USE_RAW_API=1 -DOPENTHREAD_CONFIG_ASSERT_ENABLE=0 -DOPENTHREAD_CONFIG_BACKBONE_ROUTER_DUA_NDPROXYING_ENABLE=0 -DOPENTHREAD_CONFIG_BACKBONE_ROUTER_MULTICAST_ROUTING_ENABLE=0 -DOPENTHREAD_CONFIG_ENABLE_BUILTIN_MBEDTLS=0 -DOPENTHREAD_CONFIG_ENABLE_BUILTIN_MBEDTLS_MANAGEMENT=0 -DOPENTHREAD_CONFIG_LOG_LEVEL=1 -DOPENTHREAD_CONFIG_MAC_CSL_AUTO_SYNC_ENABLE=0 -DOPENTHREAD_CONFIG_NCP_HDLC_ENABLE=1 -DOPENTHREAD_CONFIG_THREAD_VERSION=OT_THREAD_VERSION_1_1 -DOPENTHREAD_FTD=1 -DOPENTHREAD_PROJECT_CORE_CONFIG_FILE=\"openthread-core-zephyr-config.h\" -DOPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT=0 -DPACKAGE_NAME=\"OPENTHREAD\" -D_FORTIFY_SOURCE=2 -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR__=1 -Imodules/openthread/build/etc/cmake -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/etc/cmake -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/include -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/src/core -I/workdir/modules/crypto/mbedtls/mbedtls/include -I/workdir/zephyr/modules/mbedtls/configs -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/third_party/tcplp/bsdtcp -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/third_party/tcplp/lib -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/src -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/examples/platforms/zephyr -I/workdir/zephyr/include -Izephyr/include/generated -I/workdir/zephyr/soc/arm/nordic_nrf/nrf52 -I/workdir/zephyr/lib/libc/newlib/include -I/workdir/zephyr/lib/util/fnmatch/. -I/workdir/zephyr/subsys/net/lib/sockets/. -I/workdir/zephyr/subsys/net/lib/conn_mgr/. -I/workdir/zephyr/subsys/net/lib/openthread/platform/. -I/workdir/zephyr/subsys/settings/include -I/var/lib/buildkite-agent/zephyr-module-cache/modules/hal/cmsis/CMSIS/Core/Include -I/workdir/modules/hal/nordic/nrfx -I/workdir/modules/hal/nordic/nrfx/drivers/include -I/workdir/modules/hal/nordic/nrfx/mdk -I/workdir/zephyr/modules/hal_nordic/nrfx/. -I/workdir/modules/debug/segger/SEGGER -I/workdir/modules/debug/segger/Config -I/workdir/zephyr/modules/segger/. -I/workdir/modules/hal/nordic/drivers/nrf_802154/driver/include -I/workdir/modules/hal/nordic/drivers/nrf_802154/sl/include -I/workdir/modules/hal/nordic/drivers/nrf_802154/sl/sl_opensource/include -isystem /workdir/zephyr/subsys/cpp/include -isystem /var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/zephyr/../include -isystem /var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/zephyr/../examples/platforms -Wall -Wextra -Wshadow -Wno-c++14-compat -fno-exceptions -Os -fcheck-new -std=c++11 -Wno-register -fno-rtti -imacros /workdir/zephyr/twister-out/nrf52840dk_nrf52840/samples/net/sockets/echo_server/sample.net.sockets.echo_server.nrf_openthread/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfp16-format=ieee -imacros /workdir/zephyr/include/toolchain/zephyr_stdint.h -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/workdir/zephyr/samples/net/sockets/echo_server=CMAKE_SOURCE_DIR -fmacro-prefix-map=/workdir/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/workdir=WEST_TOPDIR -ffunction-sections -fdata-sections -specs=nano.specs -fno-builtin -std=c++11 -MD -MT modules/openthread/build/src/core/CMakeFiles/openthread-ftd.dir/common/random_manager.cpp.obj -MF modules/openthread/build/src/core/CMakeFiles/openthread-ftd.dir/common/random_manager.cpp.obj.d -o modules/openthread/build/src/core/CMakeFiles/openthread-ftd.dir/common/random_manager.cpp.obj -c /var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/src/core/common/random_manager.cpp
  | /var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/src/core/common/random_manager.cpp:39:10: fatal error: mbedtls/entropy_poll.h: No such file or directory
  | 39 \| #include <mbedtls/entropy_poll.h>
  | \|          ^~~~~~~~~~~~~~~~~~~~~~~~


```

microbuilder commented:
> Seems the mbedtls bump broke openthread. From CI:
> 
> ```
> FAILED: modules/openthread/build/src/core/CMakeFiles/openthread-ftd.dir/common/random_manager.cpp.obj
> --
>   | ccache /opt/toolchains/zephyr-sdk-0.13.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-g++ -DBUILD_VERSION=zephyr-v2.6.0-2909-ge0578a257278 -DKERNEL -DMBEDTLS_CONFIG_FILE=\"config-tls-generic.h\" -DNRF52840_XXAA -DNRF_802154_ACK_TIMEOUT_ENABLED=1 -DNRF_802154_CCA_CORR_LIMIT_DEFAULT=2 -DNRF_802154_CCA_CORR_THRESHOLD_DEFAULT=45 -DNRF_802154_CCA_ED_THRESHOLD_DEFAULT=45 -DNRF_802154_CCA_MODE_DEFAULT=NRF_RADIO_CCA_MODE_ED -DNRF_802154_CSMA_CA_ENABLED=1 -DNRF_802154_DELAYED_TRX_ENABLED=0 -DNRF_802154_ENCRYPTION_ENABLED=0 -DNRF_802154_FRAME_TIMESTAMP_ENABLED=0 -DNRF_802154_IE_WRITER_ENABLED=0 -DNRF_802154_IFS_ENABLED=0 -DNRF_802154_INTERNAL_RADIO_IRQ_HANDLING=0 -DNRF_802154_PENDING_EXTENDED_ADDRESSES=16 -DNRF_802154_PENDING_SHORT_ADDRESSES=16 -DNRF_802154_RX_BUFFERS=16 -DNRF_802154_SECURITY_WRITER_ENABLED=0 -DNRF_802154_TX_STARTED_NOTIFY_ENABLED=1 -DNRF_802154_USE_RAW_API=1 -DOPENTHREAD_CONFIG_ASSERT_ENABLE=0 -DOPENTHREAD_CONFIG_BACKBONE_ROUTER_DUA_NDPROXYING_ENABLE=0 -DOPENTHREAD_CONFIG_BACKBONE_ROUTER_MULTICAST_ROUTING_ENABLE=0 -DOPENTHREAD_CONFIG_ENABLE_BUILTIN_MBEDTLS=0 -DOPENTHREAD_CONFIG_ENABLE_BUILTIN_MBEDTLS_MANAGEMENT=0 -DOPENTHREAD_CONFIG_LOG_LEVEL=1 -DOPENTHREAD_CONFIG_MAC_CSL_AUTO_SYNC_ENABLE=0 -DOPENTHREAD_CONFIG_NCP_HDLC_ENABLE=1 -DOPENTHREAD_CONFIG_THREAD_VERSION=OT_THREAD_VERSION_1_1 -DOPENTHREAD_FTD=1 -DOPENTHREAD_PROJECT_CORE_CONFIG_FILE=\"openthread-core-zephyr-config.h\" -DOPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT=0 -DPACKAGE_NAME=\"OPENTHREAD\" -D_FORTIFY_SOURCE=2 -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR__=1 -Imodules/openthread/build/etc/cmake -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/etc/cmake -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/include -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/src/core -I/workdir/modules/crypto/mbedtls/mbedtls/include -I/workdir/zephyr/modules/mbedtls/configs -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/third_party/tcplp/bsdtcp -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/third_party/tcplp/lib -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/src -I/var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/examples/platforms/zephyr -I/workdir/zephyr/include -Izephyr/include/generated -I/workdir/zephyr/soc/arm/nordic_nrf/nrf52 -I/workdir/zephyr/lib/libc/newlib/include -I/workdir/zephyr/lib/util/fnmatch/. -I/workdir/zephyr/subsys/net/lib/sockets/. -I/workdir/zephyr/subsys/net/lib/conn_mgr/. -I/workdir/zephyr/subsys/net/lib/openthread/platform/. -I/workdir/zephyr/subsys/settings/include -I/var/lib/buildkite-agent/zephyr-module-cache/modules/hal/cmsis/CMSIS/Core/Include -I/workdir/modules/hal/nordic/nrfx -I/workdir/modules/hal/nordic/nrfx/drivers/include -I/workdir/modules/hal/nordic/nrfx/mdk -I/workdir/zephyr/modules/hal_nordic/nrfx/. -I/workdir/modules/debug/segger/SEGGER -I/workdir/modules/debug/segger/Config -I/workdir/zephyr/modules/segger/. -I/workdir/modules/hal/nordic/drivers/nrf_802154/driver/include -I/workdir/modules/hal/nordic/drivers/nrf_802154/sl/include -I/workdir/modules/hal/nordic/drivers/nrf_802154/sl/sl_opensource/include -isystem /workdir/zephyr/subsys/cpp/include -isystem /var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/zephyr/../include -isystem /var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/zephyr/../examples/platforms -Wall -Wextra -Wshadow -Wno-c++14-compat -fno-exceptions -Os -fcheck-new -std=c++11 -Wno-register -fno-rtti -imacros /workdir/zephyr/twister-out/nrf52840dk_nrf52840/samples/net/sockets/echo_server/sample.net.sockets.echo_server.nrf_openthread/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfp16-format=ieee -imacros /workdir/zephyr/include/toolchain/zephyr_stdint.h -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/workdir/zephyr/samples/net/sockets/echo_server=CMAKE_SOURCE_DIR -fmacro-prefix-map=/workdir/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/workdir=WEST_TOPDIR -ffunction-sections -fdata-sections -specs=nano.specs -fno-builtin -std=c++11 -MD -MT modules/openthread/build/src/core/CMakeFiles/openthread-ftd.dir/common/random_manager.cpp.obj -MF modules/openthread/build/src/core/CMakeFiles/openthread-ftd.dir/common/random_manager.cpp.obj.d -o modules/openthread/build/src/core/CMakeFiles/openthread-ftd.dir/common/random_manager.cpp.obj -c /var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/src/core/common/random_manager.cpp
>   | /var/lib/buildkite-agent/zephyr-module-cache/modules/lib/openthread/src/core/common/random_manager.cpp:39:10: fatal error: mbedtls/entropy_poll.h: No such file or directory
>   | 39 \| #include <mbedtls/entropy_poll.h>
>   | \|          ^~~~~~~~~~~~~~~~~~~~~~~~
> ```

Fixes the TF-M issues, though, thanks.

gilles-peskine-arm commented on modules/mbedtls/configs/config-tls-generic.h:
```diff
@@ -241,6 +241,7 @@
 #endif
 
 #if defined(CONFIG_MBEDTLS_MAC_SHA256_ENABLED)
+#define MBEDTLS_SHA224_C
```
This configuration enables `MBEDTLS_SHA512_C`. In Mbed TLS 2.x this enabled SHA-384, but in 3.0 this is no longer the case. So you should probably add `#define MBEDTLS_SHA384_C` as well.

gilles-peskine-arm commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -18,6 +18,20 @@ LOG_MODULE_REGISTER(net_sock_tls, CONFIG_NET_SOCKETS_LOG_LEVEL);
 #include <syscall_handler.h>
 #include <sys/fdtable.h>
 
+/* TODO: Remove all direct access to private fields.
```
`mbedtls_2to3.py` from https://github.com/ARMmbed/mbedtls/pull/4800 may help to locate all the places where a private field is needed. If there are fields that you can't avoid accessing, please raise an issue on the mbedtls project (if there isn't one already).

d3zd3z committed:
modules: mbedtls: Fix buffer size configs for 3.0

Mbed TLS 3.0 changes the configuration defines used to determine the
buffer size used for TLS.  We were still setting the old one, which was
causing Mbed TLS to revert back to the large default buffer size.  Set
both the in and out buffer sizes to match the config setting.  A future
improvement may be to separate this into two configurations within the
Zephyr config.

Signed-off-by: David Brown <david.brown@linaro.org>

SebastianBoe commented:
Rebased onto the openthread upmerge (https://github.com/zephyrproject-rtos/zephyr/pull/38497).

Bumping mbedtls_3 requires mbedtls_3 support in openthread which requires the upmerge.

Please let me know if you want this reset @d3zd3z . But then we won't be able to test mbedtls3 + openthread.

d3zd3z commented:
This is fine, and seems like a reasonable way to get all of the needed pieces together.

d3zd3z commented:
@SebastianBoe any idea why this is still failing CI on openthread?

SebastianBoe commented:
Hi, we are currently upmerging OpenThread.

When that is complete I will rebase this PR again and see if the failures are resolved.

We will no longer need to have an openthread bump in this PR when it is complete.

The latest Openthread now supports mbedtls 2 and 3.

SebastianBoe commented:
Updated manifest to use a different tf-m upmerge PR.

We need to use tf-m 1.4.1 instead of 1.4.0 due to a bugfix.

d3zd3z commented:
I'll keep an eye on this. Once everything is pointing to hashes instead of PRs in the west.yml file, I can change the commit(s) that add that, and that will get rid of the compliance test failing.

SebastianBoe commented:
Rebased onto main to get the latest openthread revision.

We no longer need to modify the openthread manifest.

Hopefully, openthread will be green in CI now.

d3zd3z commented:
In order to fix up the west.yml so this can be merged, the following will need to get merged:

- zephyrproject-rtos/mbedtls#29
- zephyrproject-rtos/mcuboot#60
- zephyrproject-rtos/trusted-firmware-m#61

galak reviewed (dismissed):
We should move the module to use the GIT SHA based tree as part of this.

d3zd3z committed:
manifest: Update dependencies for Mbed TLS 3.0

This updates Mbed TLS to 3.0 as well as trusted-firmware-m and mcuboot
to versions that also support this version.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z commented:
@galak writes:

> We should move the module to use the GIT SHA based tree as part of this.

The difficulty here is that this involves changing the directory layout of the mbed TLS tree, which likely involves modifying all of the dependencies that need it. Given that I believe this change is ready to merge, I suggest we merge this now, and make a separate effort to change the history of the mbed TLS repo to match upstream. That way we can worry about the issues of the directory change (and some Cmake complexities) separately from the effort that has gone into making this PR.
